### PR TITLE
fix(freemarker): pass class loader explicitely

### PR DESCRIPTION
**Pull request by a-cordier**
fix(freemarker): pass class loaders explicitely

This revision refactors the FreemarkerComponent to force passing a class loader explicitly when templates should be loaded from the class path.

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `3.3.1-fix-freemarker-component-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/common/gravitee-common/3.3.1-fix-freemarker-component-SNAPSHOT/gravitee-common-3.3.1-fix-freemarker-component-SNAPSHOT.zip)

### CODE REVIEW

File: src/main/java/io/gravitee/common/templating/FreeMarkerComponent.java [hunk 1]
```diff
@@ -28,143 +28,103 @@
...
+     * If path is defined (not null), templates will be looked for on the file system first, then in the class path if
+     * a class loader has been provided.
+     *
+     * If classTemplateLoaderBase is not defined (null), templates will be looked for in the root of the class path.
      */
-    public FreeMarkerComponent(Path path, String subPackagePath) {
-        var basePackagePath = sanitizeBasePackage(subPackagePath);
-        this.freeMarkerConfiguration = freemarkerConfiguration(path, basePackagePath);
+    @Builder
+    public FreeMarkerComponent(Path path, ClassLoader classLoader, String classTemplateLoaderBase) {
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/common/templating/FreeMarkerComponent.java:62]: `classLoaderTemplateBase`?

File: src/main/java/io/gravitee/common/templating/FreeMarkerComponent.java [hunk 2]
```diff
@@ -28,143 +28,103 @@
...
-     * {@link FreeMarkerComponent#FreeMarkerComponent(Path)}
+     * Creates a new FreeMarkerComponent.
+     * @param path path to a template directory on the File System
+     * @param classLoader class loader to load templates from the class path
+     * @param classTemplateLoaderBase base path to load templates from the class path
+     *
+     * If path is defined (not null), templates will be looked for on the file system first, then in the class path if
+     * a class loader has been provided.
+     *
+     * If classTemplateLoaderBase is not defined (null), templates will be looked for in the root of the class path.
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/common/templating/FreeMarkerComponent.java:59]: Does it mean that if I use the builder like this:

`FreeMarkerComponent.builder().classLoader(classLoader).build()`, then I can evaluate a template with `freemarkerComponent.generateFromTemplate("freemarker/test.tpl", ...)` --> note doesn't start with '/'

But if I use it like this `FreeMarkerComponent.builder().classLoader(classLoader).classTemplateLoaderBase("/freemarker").build()` then I use it like this `freemarkerComponent.generateFromTemplate("/test.tpl", ...)` --> note the '/' at the beginning

If so, do you think we can try to align the usage by always expecting something starting with '/'?

**Review by jhaeyaert — CHANGES_REQUESTED** [review]

File: src/main/java/io/gravitee/common/templating/FreeMarkerComponent.java (hunk 2 above)
Comment (a-cordier) [reply src/main/java/io/gravitee/common/templating/FreeMarkerComponent.java:59]: FreeMarker handles that for us. e.g.

I can use either one of
  - `classTemplateLoaderBase("freemarker/")`
  - `classTemplateLoaderBase("freemarker")`
  - `classTemplateLoaderBase("freemarker/")`
  
And get the same result with:
  - `freemarkerComponent.generateFromTemplate("/test.tpl", data)`
  
Which leaves us with a consistent behaviour

**Review by jhaeyaert — APPROVED** [review]

**Pull request merged**